File: mastermind-solver-js/api_test.mjs
import loadWasm from './mastermind.js'

const instance = await loadWasm()

const codeLength = 100
const solver = instance._Solver_ctor(6,codeLength)

const pArr1 = instance._malloc(Int32Array.BYTES_PER_ELEMENT*codeLength)
const arr1 = new Int32Array(instance.HEAP8.buffer, pArr1, codeLength)
arr1.set(Array(codeLength).fill(0))
instance._add_interaction(solver, pArr1, 2, 0)
instance._free(pArr1)

const pSecret = instance._malloc(Int32Array.BYTES_PER_ELEMENT*codeLength)
const ok = instance._solve(solver, pSecret) === 1
if(!ok) throw "Solver failed to return a solution."

const secret = new Int32Array(instance.HEAP8.buffer, pSecret, codeLength)
if(secret.filter(x => x === 0).length !== 2) throw `Result must contain exactly two zeros: ${secret}`

instance._free(pSecret)
instance._Solver_dtor(solver)
console.log("Successful")

File: mastermind-solver-cpp/test/api_test.cpp
#include <cstdio>
#include "api.h"
#include "gtest/gtest.h"

using namespace mastermind;

TEST(API, solve) {
    auto solver = Solver_ctor(6,4);

    auto guess1 = new int[]{0, 5, 5, 4};
    add_interaction(solver, guess1, 0, 1);
    delete[] guess1;

    auto guess2 = new int[]{1, 3, 0, 0};
    add_interaction(solver, guess2, 1, 2);
    delete[] guess2;

    auto guess3 = new int[]{3, 2, 0, 1};
    add_interaction(solver, guess3, 2, 1);
    delete[] guess3;

    auto secret = new int[4];
    auto ok = solve(solver, secret);

    EXPECT_TRUE(ok);
    EXPECT_EQ(std::vector<int>({2,0,0,1}), std::vector(secret, secret+4));

    delete[] secret;
    Solver_dtor(solver);
}